### PR TITLE
[EVM] Initial support of EVM dependencies

### PR DESCRIPTION
**Pull request by PavelKopyl**
[EVM] Adding LLVMLinkEVM C-API.
    Please, note this is a temporary patch. It adds initial support
    of dependencies, but it doesn't work in a general case.
    A full solution will be more sophisticated and will likely be
    implemented on the FE driver side without a need of the LLD usage.
 
 [LLD] Adding EVM architecture to LLD.

 [EVM] Adding datasize/dataoffset intrinsics.   
    This includes:
     - datasize/dataoffset intrinsics
     - EVM specific ELF relocations

# Code Review Checklist

## Requirements
- [ ] Have the requirements been met?
- [ ] Have stakeholder(s) approved the change?

## Implementation
- [ ] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified?
- [ ] Does this change add unwanted compile-time or run-time dependencies?
- [ ] Could an additional framework, API, library, or service improve the solution?
- [ ] Could we reuse part of LLVM instead of implementing the patch or a part of it?
- [ ] Is the code at the right abstraction level?
- [ ] Is the code modular enough?
- [ ] Can a better solution be found in terms of maintainability, readability, performance, or security?
- [ ] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [ ] Are there any best practices, design patterns or language-specific patterns that could substantially improve this code? 

## Logic Errors and Bugs
- [ ] Can you think of any use case in which the
code does not behave as intended?
- [ ] Can you think of any inputs or external events
that could break the code?

## Error Handling and Logging
- [ ] Is error handling done the correct way?
- [ ] Should any logging or debugging information
be added or removed?
- [ ] Are error messages user-friendly?
- [ ] Are there enough log events and are they
written in a way that allows for easy
debugging?

## Maintainability
- [ ] Is the code easy to read?
- [ ] Is the code not repeated (DRY Principle)?
- [ ] Is the code method/class not too long?

## Dependencies
- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Are there any potential impacts on other parts of the system or backward compatibility?

## Security
- [ ] Does the code introduce any security vulnerabilities?

## Performance
- [ ] Do you think this code change decreases
system performance?
- [ ] Do you see any potential to improve the
performance of the code significantly?

## Testing and Testability
- [ ] Is the code testable?
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
	- [ ] For changes to mutable state
- [ ] Do tests reasonably cover the code change (unit/integration/system tests)? 
	- [ ] Line Coverage
	- [ ] Region Coverage
	- [ ] Branch Coverage
- [ ] Are there some test cases, input or edge cases
that should be tested in addition?

## Readability
- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by
smaller methods?
- [ ] Can the readability of the code be improved by
different function, method or variable names?
- [ ] Is the code located in the right
file/folder/package?
- [ ] Do you think certain methods should be
restructured to have a more intuitive control
flow?
- [ ] Is the data flow understandable?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message
better?
- [ ] Would more comments make the code more
understandable?
- [ ] Could some comments be removed by making the code itself more readable?
- [ ] Is there any commented-out code?
- [ ] Have you run a spelling and grammar checker?

## Documentation
- [ ] Is there sufficient documentation?
- [ ] Is the ReadMe.md file up to date?

## Best Practices
- [ ] Follow Single Responsibility principle?
- [ ] Are different errors handled correctly?
- [ ] Are errors and warnings logged?
- [ ] Magic values avoided?
- [ ] No unnecessary comments?
- [ ] Minimal nesting used?

## Experts' Opinion
- [ ] Do you think a specific expert, like a security
expert or a usability expert, should look over
the code before it can be accepted?
- [ ] Will this code change impact different teams, and should they review the change as well?

### CODE REVIEW

File: lld/ELF/Arch/EVM.cpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+static uint32_t getEFlags(InputFile *file) {
+  return cast<ObjFile<ELF32LE>>(file)->getObj().getHeader().e_flags;
+}
+
+uint32_t EVM::calcEFlags() const {
+  assert(!ctx.objectFiles.empty());
+
+  const uint32_t flags = getEFlags(ctx.objectFiles[0]);
+  for (InputFile *f : ArrayRef(ctx.objectFiles).slice(1)) {
```
Comment (akiramenai) [inline lld/ELF/Arch/EVM.cpp]: Nit: Can't we use C++17 for it?

File: lld/ELF/Arch/EVM.cpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  RelExpr getRelExpr(RelType type, const Symbol &s,
+                     const uint8_t *loc) const override;
+  void relocate(uint8_t *loc, const Relocation &rel,
+                uint64_t val) const override;
+};
+} // namespace
+
+RelExpr EVM::getRelExpr(RelType type, const Symbol &s,
+                        const uint8_t *loc) const {
+  switch (type) {
```
Comment (akiramenai) [inline lld/ELF/Arch/EVM.cpp:50]: Nit: if we expect `R_EVM_DATA` to remain the only relocation we need, we could use less verbose variant of this check.

File: lld/include/lld-c/LLDAsLibraryC.h [hunk 3]
```diff
@@ -123,6 +123,32 @@ char **LLVMGetUndefinedLinkerSymbolsEraVM(LLVMMemoryBufferRef inBuffer,
  *  LLVMGetUndefinedSymbolsEraVM(). */
 void LLVMDisposeUndefinedLinkerSymbolsEraVM(char *linkerSymbolNames[],
                                             uint64_t numLinkerSymbols);
+
+/** Links the deploy and runtime ELF object files using the information about
+ *  dependencies.
+ *  \p inBuffers - array of input memory buffers with following structure:
+ *
+ *   inBuffers[0] - deploy ELF object code
+ *   inBuffers[1] - deplyed (runtime) ELF object code
```
Comment (akiramenai) [inline lld/include/lld-c/LLDAsLibraryC.h]: typo

File: lld/include/lld-c/LLDAsLibraryC.h [hunk 4]
```diff
@@ -123,6 +123,32 @@ char **LLVMGetUndefinedLinkerSymbolsEraVM(LLVMMemoryBufferRef inBuffer,
...
+ *   --------------------------
+ *   inBuffers[2] - 1-st sub-contract (final EVM bytecode)
+ *   ...
+ *   inBuffers[N] - N-st sub-contract (final EVM bytecode)
+ *
+ *  Sub contracts are optional. They should have the same ordering as in
```
Comment (akiramenai) [inline lld/include/lld-c/LLDAsLibraryC.h]: sub-contracts for consistency.

File: lld/lld-c/LLDAsLibraryC.cpp [hunk 5]
```diff
@@ -443,3 +443,199 @@ void LLVMDisposeUndefinedLinkerSymbolsEraVM(char *linkerSymbolNames[],
...
   std::free(linkerSymbolNames);
 }
+
+//----------------------------------------------------------------------------//
+
+/// This function generates a linker script for EVM architecture.
+/// \p memBufs - array of input memory buffers with following structure:
+///
+///   memBufs[0] - deploy object code
+///   memBufs[1] - deplyed object code
```
Comment (akiramenai) [inline lld/lld-c/LLDAsLibraryC.cpp]: typo

File: lld/lld-c/LLDAsLibraryC.cpp [hunk 6]
```diff
@@ -443,3 +443,199 @@ void LLVMDisposeUndefinedLinkerSymbolsEraVM(char *linkerSymbolNames[],
...
+///   --------------------------
+///   memBufs[2] - 1-st sub-contract (final EVM bytecode)
+///   ...
+///   memBufs[N] - N-st sub-contract (final EVM bytecode)
+///
+/// Sub contracts are optional. They should have the same ordering as in
```
Comment (akiramenai) [inline lld/lld-c/LLDAsLibraryC.cpp]: sub-contracts

File: lld/lld-c/LLDAsLibraryC.cpp [hunk 7]
```diff
@@ -443,3 +443,199 @@ void LLVMDisposeUndefinedLinkerSymbolsEraVM(char *linkerSymbolNames[],
...
+///       D_105_deployed(.text);
+///       __datasize_D_105_deployed = . - __dataoffset_D_105_deployed;
+///       __dataoffset_B_40 = .;
+///       __datasize_D_105 = __dataoffset_B_40 + __datasize_B_40;
+///       LONG(__dataoffset_D_105_deployed);
+///     }
+///
+/// The dot '.' denotes current location in the resulting file.
+/// The purpose of the script is to define datasize/dataoffset absolute symbols
+/// that reflect the YUL layour.
```
Comment (akiramenai) [inline lld/lld-c/LLDAsLibraryC.cpp]: typo

File: lld/lld-c/LLDAsLibraryC.cpp [hunk 8]
```diff
@@ -443,3 +443,199 @@ void LLVMDisposeUndefinedLinkerSymbolsEraVM(char *linkerSymbolNames[],
...
+  //   __datasize_B_40 = 1384;
+  std::string symDatasizeDeps;
+
+  // Contains symbols whose values are the offsets of the dependent contracts.
+  // For the example above, this contains:
+  //   __dataoffset_B_40 = .;
+  std::string symDataOffsetDeps;
+  if (memBufs.size() > 2) {
+    // Define datasize symbols for the dependent contracts. They start after
+    // {deploy, deployed} pair of the top-level contract,.i.e. at index 2.
```
Comment (akiramenai) [inline lld/lld-c/LLDAsLibraryC.cpp]: redundant dot

File: lld/lld-c/LLDAsLibraryC.cpp [hunk 9]
```diff
@@ -443,3 +443,199 @@ void LLVMDisposeUndefinedLinkerSymbolsEraVM(char *linkerSymbolNames[],
...
+    for (unsigned idx = 2; idx < bufIDs.size(); ++idx)
```
Comment (akiramenai) [inline lld/lld-c/LLDAsLibraryC.cpp]: Nit: Could we avoid recomputing `bufIDs.size()` here and below.

File: lld/lld-c/LLDAsLibraryC.cpp [hunk 10]
```diff
@@ -443,3 +443,199 @@ void LLVMDisposeUndefinedLinkerSymbolsEraVM(char *linkerSymbolNames[],
...
+      symDatasizeDeps += (dataSizePrefix + bufIDs[idx] + " = " +
+                          Twine(LLVMGetBufferSize(memBufs[idx])) + ";\n")
+                             .str();
+
+    symDataOffsetDeps = (dataOffsetPrefix + bufIDs[2] + " = .;\n").str();
+    for (unsigned idx = 3; idx < bufIDs.size(); ++idx) {
```
Comment (akiramenai) [inline lld/lld-c/LLDAsLibraryC.cpp]: redundant brackets

File: lld/lld-c/LLDAsLibraryC.cpp [hunk 11]
```diff
@@ -443,3 +443,199 @@ void LLVMDisposeUndefinedLinkerSymbolsEraVM(char *linkerSymbolNames[],
...
+      symDataOffsetDeps +=
+          (dataOffsetPrefix + bufIDs[idx] + " = " + dataOffsetPrefix +
+           bufIDs[idx - 1] + " + " + dataSizePrefix + bufIDs[idx - 1] + ";\n")
+              .str();
+    }
+  }
+
+  // Contains a symbols whose value is the total size of the top-level contract
```
Comment (akiramenai) [inline lld/lld-c/LLDAsLibraryC.cpp]: remove `a`

File: llvm/lib/Target/EVM/EVMAsmPrinter.cpp [hunk 12]
```diff
@@ -70,12 +76,43 @@ void EVMAsmPrinter::emitFunctionEntryLabel() {
...
+  // Check terminators in more details. After the stackification we may have the
+  // following case:
+
+  //   PUSH_LABEL %bb.11
+  //   JUMPI %bb.11, %27:gpr..
+  //   PUSH_LABEL %bb.9
+  //   JUMP %bb.9
+  // bb.11:
+  //
+  // The default implementation checks only the first terminator, JUMP %bb.9
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMAsmPrinter.cpp]: > the first terminator

the last one if to read instructions as they follow.

File: llvm/lib/Target/EVM/EVMAsmPrinter.cpp [hunk 13]
```diff
@@ -70,12 +76,43 @@ void EVMAsmPrinter::emitFunctionEntryLabel() {
 void EVMAsmPrinter::emitInstruction(const MachineInstr *MI) {
   EVMMCInstLower MCInstLowering(OutContext, *this, VRegMapping,
                                 MF->getRegInfo());
-
   MCInst TmpInst;
   MCInstLowering.Lower(MI, TmpInst);
   EmitToStreamer(*OutStreamer, TmpInst);
 }
 
+bool EVMAsmPrinter::isBlockOnlyReachableByFallthrough(
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMAsmPrinter.cpp:102]: Can we add a test?

File: llvm/lib/Target/EVM/MCTargetDesc/EVMAsmBackend.cpp [hunk 14]
```diff
@@ -138,7 +138,9 @@ bool EVMAsmBackend::evaluateTargetFixup(const MCAssembler &Asm,
   Value = Target.getConstant();
   if (const MCSymbolRefExpr *A = Target.getSymA()) {
     const MCSymbol &Sym = A->getSymbol();
-    assert(Sym.isDefined());
+    if (!Sym.isDefined())
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/MCTargetDesc/EVMAsmBackend.cpp]: Doesn't seem to be relevant to the commit introducing the change. Perhaps, it's better to extract to a separate fixup.

**Review by akiramenai — CHANGES_REQUESTED** [review]
I requested changes mostly because of a missing test. The rest are nitpicks.
Could we also use infinitives in the commit titles and avoid dots at the end?
I also recommend elaborating that the last change is required because EVM marks jump locations, so terminator instructions aren't adjacent. To make it a bit more EVM-novice friendly.

File: lld/ELF/Arch/EVM.cpp (hunk 1 above)
Comment (PavelKopyl) [reply lld/ELF/Arch/EVM.cpp]: Sure, updated.

File: lld/ELF/Arch/EVM.cpp (hunk 2 above)
Comment (PavelKopyl) [reply lld/ELF/Arch/EVM.cpp:50]: I'd like to leave it as it, because we will add more relocations here.

File: lld/include/lld-c/LLDAsLibraryC.h (hunk 3 above)
Comment (PavelKopyl) [reply lld/include/lld-c/LLDAsLibraryC.h]: Fixed.

File: lld/include/lld-c/LLDAsLibraryC.h (hunk 4 above)
Comment (PavelKopyl) [reply lld/include/lld-c/LLDAsLibraryC.h]: Fixed.

File: llvm/lib/Target/EVM/EVMAsmPrinter.cpp (hunk 13 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMAsmPrinter.cpp:102]: I added a test. Please, note I've changed the implementation. Not it just always returns false for simplicity. I do not sees any drawbacks here. I moved instruction stackification to the early stage (EVMStackify), so now it's actually not possible to analyze BB terminators here (and this has no much sense).

File: llvm/lib/Target/EVM/MCTargetDesc/EVMAsmBackend.cpp (hunk 14 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/MCTargetDesc/EVMAsmBackend.cpp]: Removed.

File: lld/lld-c/LLDAsLibraryC.cpp (hunk 9 above)
Comment (PavelKopyl) [reply lld/lld-c/LLDAsLibraryC.cpp]: Moved it to a variable.

File: llvm/lib/Target/EVM/EVMAsmPrinter.cpp (hunk 12 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMAsmPrinter.cpp]: I changed the description.

File: lld/ELF/Arch/EVM.cpp [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+static uint32_t getEFlags(InputFile *file) {
+  return cast<ObjFile<ELF32LE>>(file)->getObj().getHeader().e_flags;
+}
+
+uint32_t EVM::calcEFlags() const {
+  assert(!ctx.objectFiles.empty());
+
+  const uint32_t flags = getEFlags(ctx.objectFiles[0]);
+  if (auto it = std::find_if_not(ctx.objectFiles.begin(), ctx.objectFiles.end(),
+                                 [&flags](InputFile *f) {
```
Comment (akiramenai) [inline lld/ELF/Arch/EVM.cpp]: It's better to pass `flags` by value.

File: lld/ELF/Arch/EVM.cpp [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+                                   const uint32_t objFlags = getEFlags(f);
```
Comment (akiramenai) [inline lld/ELF/Arch/EVM.cpp]: Nit: You only use it once. `getEFlags(f) == flags`?

File: lld/ELF/Arch/EVM.cpp [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// EVM is a stack-based virtual machine with a word size of 256 bits intendent
+// for execution of smart contracts in Ethereum blockchain environment.
+//
+// Since it is a baremetal programming, there's usually no loader to load
```
Comment (akiramenai) [inline lld/ELF/Arch/EVM.cpp]: remove `a`

File: llvm/lib/Target/EVM/EVMInstrInfo.cpp [hunk 18]
```diff
@@ -65,6 +66,12 @@ bool EVMInstrInfo::analyzeBranch(MachineBasicBlock &MBB,
   FBB = nullptr;
   Cond.clear();
 
+  const auto *MFI = MBB.getParent()->getInfo<EVMMachineFunctionInfo>();
+  if (MFI->getIsStackified()) {
+    LLVM_DEBUG(dbgs() << "Can't analyze terminators in a stackified code");
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMInstrInfo.cpp]: remove `a`

File: llvm/lib/Target/EVM/EVMMCInstLower.cpp [hunk 19]
```diff
@@ -27,30 +27,27 @@ using namespace llvm;
...
-  unsigned RegOpcode = OutMI.getOpcode();
-  if (RegOpcode == EVM::PUSH8_LABEL) {
-    // Replace PUSH8_LABEL with PUSH8_S opcode.
-    OutMI.setOpcode(EVM::PUSH8_S);
-  } else {
-    unsigned StackOpcode = EVM::getStackOpcode(RegOpcode);
-    OutMI.setOpcode(StackOpcode);
+#ifndef NDEBUG
+  // Check there are no register operands.
+  for (unsigned I = 0; I < OutMI.getNumOperands(); ++I) {
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMMCInstLower.cpp]: `all` / `any` algorithm

File: llvm/lib/Target/EVM/EVMStackify.cpp [hunk 20]
```diff
@@ -1013,6 +1037,29 @@ void StackModel::postProcess() {
...
+      stackifyInstruction(&MI);
+
+  auto *MFI = MF->getInfo<EVMMachineFunctionInfo>();
+  MFI->setIsStackified();
+
+  // In a stackified code register liveness has no meaning.
+  MachineRegisterInfo &MRI = MF->getRegInfo();
+  MRI.invalidateLiveness();
+
+  for (MachineBasicBlock &MBB : *MF) {
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMStackify.cpp:1065]: Bundle usage is to be explained somewhere in code.

File: llvm/lib/Target/EVM/EVMTargetObjectFile.cpp [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+#include "EVMTargetObjectFile.h"
+
+using namespace llvm;
+
+// Code sections need to be aligned on 1, otherwise linker will add padding
+// between .text sections of the object files being linked.
+unsigned EVMELFTargetObjectFile::getTextSectionAlignment() const { return 1; }
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMTargetObjectFile.cpp]: Can we inline and define in the header?

**Review by akiramenai — APPROVED** [review]
LGTM, except for nitpicks.

File: lld/ELF/Arch/EVM.cpp (hunk 15 above)
Comment (PavelKopyl) [reply lld/ELF/Arch/EVM.cpp]: Done.

File: lld/ELF/Arch/EVM.cpp (hunk 16 above)
Comment (PavelKopyl) [reply lld/ELF/Arch/EVM.cpp]: Done.

File: lld/ELF/Arch/EVM.cpp (hunk 17 above)
Comment (PavelKopyl) [reply lld/ELF/Arch/EVM.cpp]: Done.

File: llvm/lib/Target/EVM/EVMInstrInfo.cpp (hunk 18 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMInstrInfo.cpp]: Done.

File: llvm/lib/Target/EVM/EVMMCInstLower.cpp (hunk 19 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMMCInstLower.cpp]: Done.

File: llvm/lib/Target/EVM/EVMStackify.cpp (hunk 20 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMStackify.cpp:1065]: Added the description.

File: llvm/lib/Target/EVM/EVMTargetObjectFile.cpp (hunk 21 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMTargetObjectFile.cpp]: Done.

File: llvm/tools/llvm-readobj/ELFDumper.cpp [hunk 22]
```diff
@@ -1261,6 +1261,9 @@ const EnumEntry<unsigned> ElfMachineType[] = {
   // EraVM local begin
   ENUM_ENT(EM_ERAVM,         "EraVM"),
   // EraVM local end
+  // EVM local begin
+  ENUM_ENT(EM_EVM,           "Ethereum Virtual machine"),
```
Comment (hedgar2017) [inline llvm/tools/llvm-readobj/ELFDumper.cpp]: Please make consistent with the `EraVM` above.

File: llvm/test/CodeGen/EVM/fallthrough.mir [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+  bb.0:
+      JUMPDEST_S
+      PUSH_LABEL %bb.10 {
+        JUMPI_S
+      }
+      PUSH_LABEL %bb.13 {
+        JUMP_S
+      }
+
+  bb.10:
```
Comment (hedgar2017) [inline llvm/test/CodeGen/EVM/fallthrough.mir:32]: Wow. Is it kinda EVM assembly?

**Review by hedgar2017 — CHANGES_REQUESTED** [review]

File: llvm/test/CodeGen/EVM/fallthrough.mir (hunk 23 above)
Comment (PavelKopyl) [reply llvm/test/CodeGen/EVM/fallthrough.mir:32]: No quite. This is LLVM machine IR representation, but it's quite close to the assembler.

File: llvm/tools/llvm-readobj/ELFDumper.cpp (hunk 22 above)
Comment (PavelKopyl) [reply llvm/tools/llvm-readobj/ELFDumper.cpp]: Done.

**Review by hedgar2017 — APPROVED** [review]
